src/borg/repository.py

```diff
@@ -1184,13 +1184,7 @@ def put(self, id, data, wait=True):
         except KeyError:
             pass
         else:
-            self.segments[segment] -= 1
-            size = self.io.read(segment, offset, id, read_data=False)
-            self.storage_quota_use -= size
-            self.compact[segment] += size
-            segment, size = self.io.write_delete(id)
-            self.compact[segment] += size
-            self.segments.setdefault(segment, 0)
+            self._delete(id, segment, offset)
```

Comment:
It might be worth including a comment here indicating that the shadow_index will not get updated without performing an explicit delete here.

I'm not sure what to make of how borg's repository layer allows a PUT to shadow another PUT, but currently never *does* that (instead using DELETE + PUT). I'm not sure if `PUT A, PUT A` should be considered a sequence in need of repair.

Reply:
added a comment.

seems like put del put is also needed for other bookkeeping.
e.g. imagine borg recreate with recompress - data size will usually change here, but chunkid will stay same.